### PR TITLE
Highscores, joystick scrolling, play count tracking, OnlineLibrary features

### PR DESCRIPTION
**Pull request by hankedan000**
- Highscore tracking
- Play count tracking
- Playlist selection (all songs, most played, recently added)
- OnlineLibrary can search for maps by a song author
- OnlineLibrary automatically loads pages when scrolling
- Joystick scrolling support
- Hides track while paused (cubes could block the menu otherwise)

### CODE REVIEW

File: OQ_Toolkit/OQ_ARVRController/scripts/Feature_UIRayCast.gd [hunk 1]
```diff
@@ -32,7 +32,7 @@ func _set_raycast_transform():
 		else:
 			ui_raycast_position.transform.basis = Basis(Vector3(deg2rad(-90),0,0));
 	else:
-		ui_raycast_position.transform.basis = Basis();
```
Comment (hankedan000) [inline OQ_Toolkit/OQ_ARVRController/scripts/Feature_UIRayCast.gd:35]: @NeoSpark314 this was causing the raycast node to get offset when switching from hand tracking to controller model. Thoughts on PR this to the toolkit baseline?

File: game/BeepSaber_Game.gd [hunk 2]
```diff
@@ -2,6 +2,16 @@
 #
 extends Spatial
 
+# TODO add better state management
```
Comment (hankedan000) [inline game/BeepSaber_Game.gd]: I should probably elaborate on this...
The code makes a bunch of stateful assumptions everywhere (ie. node visibility).
I started to introduce the concept of game states, but then never followed through with the full refactor.
Thoughts?

File: game/BeepSaber_Game.gd [hunk 3]
```diff
@@ -651,4 +760,34 @@ func _on_Pause_Panel_continue_button():
 	continue_map()
 
 
+# FIXME this doesn't always seem to get called :(
```
Comment (hankedan000) [inline game/BeepSaber_Game.gd]: hmmm not sure why this was an issue... I could just save the highscore file each time we make an update. That's probably not that costly.

**Comment by leandrodreamer** [comment]
great additions so far, but i think i found a bug, i was unable to get out of the score screen, buttons stay disabled

https://user-images.githubusercontent.com/26260718/121462633-80d99880-c976-11eb-8ecf-385a1d153040.mp4

also on that screen i felt some frame drops, but it may be because of recording, not sure

File: OQ_Toolkit/OQ_ARVRController/scripts/Feature_UIRayCast.gd (hunk 1 above)
Comment (NeoSpark314) [reply OQ_Toolkit/OQ_ARVRController/scripts/Feature_UIRayCast.gd:35]: Resetting also the position and not only rotation looks correct here. You are right; we should fix this also in the toolkit main repo.

**Comment by hankedan000** [comment]
Thanks for testing! I pushed a commit that reconnects the text input signal from the keyboard. This must have gotten dropped in a bad conflict resolution.
I didn't pick up on this because once you have existing names in the highscore table, you can pick from those name instead of entering your name via the keyboard.

As for the frame drop, I think I've noticed that it's related to the keyboard UI. I tried applying some simple flat themes to the keyboard in another project and it also caused frame drop. I'll take a look around to see if it's anything else I could have introduced though.

You comment also reminds me that I wanted to add some sort of notification that you got a new high score. Maybe some flashy test that pop up on the score board, or even a little giggle. I'll probably work on adding that before closing the PR.

File: game/BeepSaber_Game.gd (hunk 2 above)
Comment (NeoSpark314) [reply game/BeepSaber_Game.gd]: Fine for me to keep it like that for now and address this in the future. But maybe you add the information from your github comment also in the comment inside the code; I think it makes it easier for others to understand if they see it in the future without the PR.

File: game/BeepSaber_Game.gd (hunk 3 above)
Comment (NeoSpark314) [reply game/BeepSaber_Game.gd]: yes; saving for each update should not be a problem

**Comment by NeoSpark314** [comment]
These are some great additions :-); very nice. I will give it a test later today; the PR is already quite big; so I think it would be better to merge it without major changes and add the highscore notification in a separate PR afterwards. what do you think?

**Comment by hankedan000** [comment]
Thanks! Sounds good to me.

File: OQ_Toolkit/OQ_ARVRController/scripts/Feature_UIRayCast.gd (hunk 1 above)
Comment (hankedan000) [reply OQ_Toolkit/OQ_ARVRController/scripts/Feature_UIRayCast.gd:35]: PR for toolkit repo is here.
https://github.com/NeoSpark314/godot_oculus_quest_toolkit/pull/72

**Pull request merged**